Allow changing video resolution at runtime + more

ZenithMDC opened the pull request:
![evo](https://github.com/user-attachments/assets/6460bc9a-4810-4452-bb9b-d003b524d995)

**NOTE: Only tested on Windows 10. I don't have access to other OSes, but hopefully it behaves the same way on them via SDL.**

New config options: `Video.CenterWindow` `Video.DisplayFPS` `Video.DisplayFPSDivisor` `Game.TickRateDivisorOverride`
I must admit that a centered window at native resolution, but not in fullscreen, results in what looks very much like fullscreen, at least on Windows 10. The taskbar disappears and everything. I don't really like that behavior and it feels misleading when Alt+Enter appears to do "nothing". I wonder how it behaves on other OSes?

EDIT: So, this means that having `Center Window` enabled and `Full Screen` disabled at native resolution is essentially borderless windowed mode. On Windows 10, the desktop compositor is **NOT** bypassed in this configuration, but in _actual_ fullscreen mode, it is bypassed (for both borderless and exclusive modes). Anyway, this was completely unintentionally implemented on my part, but some people might like this configuration more than the other ones.

### Implementation
I took the liberty to rearrange the more frequently-accessed / sought-after video options to the top of the list. Since video display modes vary per device, during `videoInit()`, I create a heap-allocated array of unique resolutions for every video mode, which are later cycled through in a dropdown menu. I made sure to free it at exit, even though that doesn't really matter too much. Since the window can be arbitrarily resized, I made sure to accommodate for that with the special "Custom" resolution, which basically just means that the current window resolution doesn't match any of the known resolutions in the array.

Variables which store the state of fullscreen, fullscreen-exclusive, resolution, and maximize are now synchronized with the state of the window. They will be written to the ini as modified at runtime. I feel like they are more intuitive this way. For instance, if I change the resolution in-game, I'd like the resolution to be the same after I exit and launch the game again, rather than the ini being the sole place where modification is allowed. I should note that custom resolutions aren't written to the ini. Whatever the last known (as in, belonging to the array) resolution was is what will be written, in that case.

### Caveats
Since it leverages the existing system in place for resizing the window, it carries the same caveats, though only one is really obvious: the blur framebuffer is blanked if the resolution changes vertically. Horizontal changes result in warping of that buffer. Cycling the pause menu on and off will take a new screenshot and restore the blur buffer. A warped blur buffer can only be fixed by blanking the blur buffer with a vertical resolution change, followed by pause menu cycling.

I was thinking of a more elegant solution to this problem, which would probably involve pushing a special menu mode for resolution changes, which, in reality, would not render any menu but instead show a clean view of the background for a screenshot to be taken, then reapply the blur buffer and pop said menu. I imagine the same solution can be used for arbitrary resizes (like dragging the window) while the pause menu is open.

### Future Work
~I have aspirations to make the rest of the ini-exclusive video options runtime-modifiable, in a way that is user-friendly. I'll have to think about how best to abstract the framerate-related options, since they're a bit technical as is. Having the ability to toggle the current FPS in-game would be nice, too.~

EDIT: All done. Can't think of anything else I'd like to add. I deviated from my plan, here and there. I dropped `Remember Window Position`, because I don't think it would be particularly useful, and `Framebuffer Effects`, because it seems like a setting that should be enabled via querying system capabilities/features rather than by a player, much like `Allow HiDPI` should be. Btw, I found that having a `Tickrate Divisor` of 1 occasionally results in unexpected framerate pacing, so I actually prefer it overridden to 0. Here's a composite image showing the full menu:

![evo_final](https://github.com/user-attachments/assets/1befa37e-2b3c-417e-ab97-027371b27e24)

As always, if you want me to change something, let me know.

ZenithMDC committed:
Allow changing video resolution at runtime + more

Fullscreen exclusive mode is togglable at runtime, as well, as a
dropdown menu with "borderless" and "exclusive" options. All variables
pertaining to video resolution, exclusive mode, and maximized state are
now synchronized. For instance, Alt+Enter will affect the state of the
fullscreen video option and maximizing the window via the title bar will
update the maximize window option accordingly.

ZenithMDC committed:
Add video option to center the window

fgsfdsfgs commented:
There are some reasons why I haven't exposed the framerate cap in the options. Mostly two:
* The way it's implemented might be confusing for the end user ("what the fuck is a tickrate divisor?"), but that is also kind of necessary to support all possible combinations of vsync/fps cap/etc, because various setups can have different frame pacing issues. This necessitates some sort of unified option that sets both the divisor and the framerate limit, but that can also be difficult to get right (e.g. what should happen when vsync is on?).
* High FPS modes are still buggy and I'd actually not recommend playing the game at over 120 FPS.

As for the PR itself, you should probably avoid exposing Fast3D headers to the game code where possible. This can be achieved by providing thin wrappers for mode listing, e.g.:

video.h
```
typedef struct {
	s32 width;
	s32 height;
} videomode;

s32 videoGetNumVideoModes(void);
s32 videoGetVideoMode(videomode *out, const s32 index);
```

video.c
```
s32 videoGetNumVideoModes(void)
{
	return wmAPI->get_num_video_modes();
}

s32 videoGetVideoMode(videomode *out, const s32 index)
{
	return wmAPI->get_video_mode(index, &out->width, &out->height);
}
```

gfx_sdl2.cpp
```
int gfx_sdl_get_num_video_modes(void) {
    return SDL_GetNumDisplayModes(0);
}

int gfx_sdl_get_video_mode(int modenum, int *out_w, int *out_h) {
    SDL_DisplayMode sdlmode;
    if (SDL_GetDisplayMode(0, modenum, &sdlmode)) {
        *out_w = sdlmode.w;
        *out_h = sdlmode.h;
        return 1;
    }
    return 0;
}
```

There's no need to allocate memory for them either. Well, maybe if you want to cache them in video.c, but that shouldn't be necessary.

ZenithMDC commented:
It seems the framerate options are more involved than I had originally thought. Thank you both for your insights. I'll have more to say about that later, but, at the moment, I'm pretty exhausted.

Hmm... I can't just use the raw output of `SDL_GetDisplayMode()` in conjunction with `SDL_GetNumDisplayModes()`, since that results in "duplicates", or really, modes with different refresh rates but the same resolutions. That `gfx_modes.h` file was one that I added because I needed to expose the struct within it to both the Fast3D code and the game code. Although, now that I think about it, did it even need to be a struct at all? It really behaves like an array of char arrays of size 10. Hmm... well, this is a bit embarrassing. Anyway, that should be an easy change and then I don't need `gfx_modes.h` at all.

The modes array is formatted like so:
```c
{"Custom", "3840x2160", "2560x1600", "2560x1440", "2048x1536", ...}
```
I could alternatively use a fixed-size array, but, I don't want to end up causing someone with, like, 50 unique display modes to not have them all be available. Also, when it comes to selecting the resolution from the dropdown menu, do the strings really _not_ need to be cached?

ZenithMDC commented:
Okay, I see. This solves the `gfx_modes.h` issue entirely. That _is_ what the issue is about and I'm not just obsessing over `gfx_modes.h` like a crazed individual, I hope? 😂 About the conversion: doesn't that have to happen at some point, regardless? Or maybe I'm missing something? But I like your solution more overall. It's much cleaner, anyway. I'll take a closer look tomorrow.

fgsfdsfgs commented:
> About the conversion: doesn't that have to happen at some point, regardless?
You only need to convert them to string when actually displaying them in the menu, because everywhere else in the code will just need the number. It just makes it easier and makes more sense to me, as you don't have to `sscanf()`/`strol()` shit just to get the actual width and height. Not to mention it stores less data.

> That is what the issue is about and I'm not just obsessing over gfx_modes.h like a crazed individual, I hope?
I just want Fast3D to remain as encapsulated as possible because it is cleaner this way and theoretically makes changing renderers less painful. It makes more sense to me when the actual display mode retrieval is on the backend (gfx_sdl2), while video.c deals with filtering/sorting/giving them to the user.

ZenithMDC committed:
Cache ints instead of strings for video modes

ZenithMDC commented:
Okay, after implementing, for the most part, what you sent me, debugging and testing everything again to make sure things are still working correctly, evicting the string cache and its accomplice, `strtol()`, and making sure things are encapsulated as they should be, we are back to, basically — where we started. I've made some changes, too, like that `gfx_sdl2.cpp` wrapper you provided not accounting for the fact that SDL returns 0 as successful, among other things.

I opted to keep the "Custom" message for unknown video resolutions because I feel that it works better when considering that just inserting whatever the current video mode is into index 0 can result in a duplicate entry or down-right confusion when a smaller value is above a list of descending values. "Custom", on the other hand, looks special because it _is_ special. But, admittedly, it can be another source of confusion if someone is expecting to be able to enter a custom resolution once selected.

fgsfdsfgs commented:
>  the fact that SDL returns 0 as successful, among other things.

> I opted to keep the "Custom" message for unknown video resolutions

Yes, I just provided the code as a janky example for how to manage the video mode list, this is all fine.

I think we don't have to actually add a way to enter a custom resolution from the menu because someone can just edit the INI for that. Most people will probably play in borderless fullscreen anyway.

ZenithMDC commented:
> I think we don't have to actually add a way to enter a custom resolution from the menu because someone can just edit the INI for that. Most people will probably play in borderless fullscreen anyway.

Yeah, I'm not planning on incorporating that functionality.

ZenithMDC commented:
Okay, on the topic of framerate:

`vidFramerateLimit` affects the video backend's framerate whereas `g_TickRateDiv` affects game's tickrate (duh lol), or, in other words, the interval between actions in the game. The N64 version of the game always has `g_TickRateDiv` set to 1, or really, the variable it is an alias for, `g_Vars.mininc60`, so I imagine it is best to leave it at 1 whenever `vidFramerateLimit` doesn't exceed 60 or is not disabled (0). If `vidFramerateLimit` is anything else, then `g_TickRateDiv` needs to be 0, because otherwise, it would prevent the framerate from exceeding 60.

so, for example something like:
```c
if (g_TickRateDiv < 2) {
    g_TickRateDiv = (vidFramerateLimit == 0 || vidFramerateLimit > 60) ? 0 : 1;
}
```

the `g_TickRateDiv < 2` check is there in case someone wants to set `TickRateDivisor` to a specific value in the ini.

I imagine setting `g_TickRateDiv` to anything other than 1 causes gameplay differences, though, and, to be honest, I'm not convinced of the usefulness of setting `g_TickRateDiv` above 1, not to mention it's really too coarse to be useful as a general frame limiter for the video menu. I was reading elsewhere (https://github.com/fgsfdsfgs/perfect_dark/issues/343#issuecomment-2058523253) that there are goals to interpolate framerates above 60, in order to prevent bugs. And, once that is implemented, would it actually be better to just leave `g_TickRateDiv` at 1 always?

One thing I don't like, at the moment, is that setting either `vidFramerateLimit` too low or `g_TickRateDiv` too high results in audio buffer underruns and input drops. I'm not sure it's possible to emulate the slower framerate of the N64 version without these issues presenting themselves, as things currently are. 

Oh, and about vsync: vsync is king. If it's enabled, then `vidFramerateLimit` and `g_TickRateDiv` are overridden. However, it's important to note that not everyone has a monitor that refreshes at 60Hz. Btw, I read in the Wiki that vsync set "2+ - sync every N frames." Is this actually true? I looked at the code and didn't see it implemented that way. The backend `SDL_GL_SetSwapInterval()` makes no mention of it, either, in its documentation.

As for how I'd like things to be laid out in the video menu: I think it's best to just keep things simple. I was thinking of making the framerate option a slider, from 0 to, I suppose, 240. I'd disable the slider if vsync is enabled, just like I currently do for the resolution dropdown menu if `vidFullscreen` is enabled and `vidFullscreenExclusive` is 0 (aka borderless). For the time being, if vsync is disabled and someone attempts to set the framerate slider to 0, it will automatically be set to 240, just as things are currently implemented. That might seem peculiar to them, but I don't want to circumvent a feature that keeps the game, at least, playable.

So yeah. If I have a totally screwed up understanding of how this works, please let me know. 😂 I'm a bit skeptical of my own conclusion since it seems to contradict what has already been mentioned here about the framerate, particularly about `g_TickRateDiv`. Is all, really.

fgsfdsfgs commented:
Yes, it would result in less bugs if everything above 60 FPS was interpolated like some other ports do. Arguably "true" high FPS is like what the port has right now is better since there's no interpolation delay, but it does result in more issues. For example, since the game uses a 240Hz timer, Combat Boost will start breaking at over 120 FPS since it slows down time 2x. So either we need to fix this somehow or fall back to interpolation, which has its own issues.

The reason you can set the tickrate divisor above 1 is "why not". Maybe you want to hard lock the game to 30FPS and see what happens.

Something like what you're suggesting can be implemented, but you need to figure out how it will interact with vsync. There's no reliable way to determine your display's refresh rate (SDL's display info functions sometimes return bullshit and sometimes you have G-Sync, etc), so it's hard to determine what framerate the game will actually run at if vsync is enabled. If you rely on `vidFramerateLimit`, this will imply that the port will always cap framerate on the CPU side even if vsync is on, which might be undesirable, as the extra sleeping can result in shit frame pacing on some setups. 

"Sync every N frames" is a somewhat confusing way to say "it will wait for N vblanks every SwapWindow call instead for 1 vblank". Meaning that if you set it to 2 and your display runs at 60Hz, the game will lock to 30FPS. If your monitor runs at 150Hz, the game will lock to 75FPS.

Both `vidFramerateLimit` and vsync stack on top of the already existing tickrate system. They don't care about each other, other than in the very specific case of "both vsync and tickrate divisor are set to 0", in which case the game will auto-cap to 240FPS using `vidFramerateLimit`. Vsync does not "override" the other caps, it just happens at the same time. This means that you can enable vsync on your 240Hz monitor and still cap the framerate to 60 if you so desire.

Proper frame pacing is a very difficult problem, at least when you don't have swap chains. Thus all the options you can screw with, so you could possibly find a solution even on the most weird-ass setup.

ZenithMDC commented:
Good points on monitor refresh rates being inconsistent and difficult to deal with, especially the variable refresh ones. Not sure how to deal with that, at the moment. I'll think of something eventually, I suppose.

> "Sync every N frames" is a somewhat confusing way to say "it will wait for N vblanks every SwapWindow call instead for 1 vblank". Meaning that if you set it to 2 and your display runs at 60Hz, the game will lock to 30FPS. If your monitor runs at 150Hz, the game will lock to 75FPS.

Oh yeah, I know what it means, but I was actually confused about how it was actually being implemented in the code. I couldn't find where it syncs every N frames.

> Vsync does not "override" the other caps, it just happens at the same time.

For some reason, I thought they were mutually exclusive. After checking the code again, they indeed are not. There is a confusing comment about `set_target_fps()` being disabled because vsync is on, but, well, it's just a comment.

ZenithMDC commented:
I was thinking about the layout of the Extended Video Options again and came up with this:

```
Full Screen			[ ]
Full Screen Mode		<Borderless|Exclusive>
Resolution			<Custom|3840x2160|1920x1080|...>
Maximize Window			[ ]
Center Window			[ ]
Remember Window Position	[ ]
HUD Centering			<None|4:3|Wide>
-------------------------
Vsync				<Off|Interval|Adaptive>
Vsync Interval			|1-----------10|
Framerate Limit			|0----------240|
Display FPS			[ ]
-------------------------
Anti-aliasing (MSAA)		<Off|2x|4x|8x|16x>
Texture Filtering		<Nearest|Bilinear|Three Point>
GUI Texture Filtering		[ ]
-------------------------
GE64-style Muzzle Flashes	[ ]
Explosion Shake			|0-----------20|
-------------------------
Allow HiDPI			[ ]
Framebuffer Effects		[ ]
Tickrate Divisor		|0-----------10|
-------------------------
Back
```

It is organized into five categories of descending "importance/popularity":
* Window/Positioning Options
* Framerate Options
* Filtering Options
* Miscellaneous Options
* Advanced Options

`[ ]` denotes a checkbox.
`< >` denotes a dropmenu.
`| |` denotes a slider.

`Remember Window Position` and `Display FPS` aren't implemented yet. I added `Tickrate Divisor` to the advanced category so people can actually set it to something other than 0 or 1 if they _really_ want to, but it's now managed under the hood if set to either 0 or 1. It is technically possible to eliminate `Full Screen Mode` and make `Full Screen` a dropdown menu with its respective options, but with `Off` prepended to it. However, I feel that this would hurt its usability, especially borderless fullscreen mode, which disables `Resolution` if `Full Screen` is enabled. It might not be as obvious to people that they need to switch the mode to something other than `Borderless` to access `Resolution` again, compared to seeing a separate `Full Screen` checkbox option. There are some additional reasons for keeping these separate options, pertaining to windowed mode, as well as the fact that they (`vidFullscreen` and `vidFullscreenExclusive`) are separate variables in the code. `Vsync Interval` is similarly disabled if `Vsync` is not set to `Interval`, but doesn't carry the caveats of `Full Screen`.

Another thing to note: this no longer fits on a single screen and needs to be scrolled. I wanted to try to keep it all visible without the need for scrolling. Oh well. At least most of the important options will be visible without scrolling. Alternatively, I suppose it would be possible to make it so the player needs to select from one of those five aforementioned categories first, but I don't want to do that since I feel like that would become a nuisance.

Oh yes, and I need to make sure that some of these options aren't available if the hardware doesn't support them.

ZenithMDC commented:
I put `Allow HiDPI` in the Advanced category for a reason. Basically, anything under there shouldn't be tinkered with unless the player knows exactly what they're doing with those options. Of course, the concept of these categories is exactly that — just concepts. They aren't going to be called that in the actual game. So, hopefully, seeing `Allow HiDPI` next to `Framebuffer Effects` and `Tickrate Divisor` will keep the lay player away. But, idk.

fgsfdsfgs commented:
Pretty sure changing HiDPI settings requires recreating the window because it's a window flag + hint, which we don't do. I'd leave it in the ini only.

ZenithMDC commented:
After implementing and testing Vsync options, it's become apparent that my proposal is inadequate. Sliders can be disabled but don't have the appearance of a disabled menu item. Moving on. I'm currently looking at the MP Limits menu for inspiration. It seems one can add labels to specific slider values. Might be interesting...

ZenithMDC committed:
Allow changing vsync and framerate limit at runtime

Also implements a new menu item flag: MENUITEMFLAG_SLIDER_DEFERRED,
which only calls the MENUOP_GETSLIDER callback on entering dimmed mode,
storing its value for later ticks, and only calls the MENUOP_SET
callback on leaving dimmed mode. This was necessary to prevent the
framerate limit slider from immediately affecting the framerate, leading
to an annoying slow down as the value was decremented to 0.

ZenithMDC commented:
Okay, I got `Vsync` and `Framerate Limit` implemented now. I decided to get rid of `Vsync Interval`. It was easier to just wrap everything into a single dropmenu item. There was an annoying issue implementing `Framerate Limit`, but I came up with what I think is a pretty good solution, detailed below:

I implemented a new menu item flag: MENUITEMFLAG_SLIDER_DEFERRED, which only calls the MENUOP_GETSLIDER callback on entering dimmed mode, storing its value for later ticks, and only calls the MENUOP_SET callback on leaving dimmed mode. This was necessary to prevent the framerate limit slider from immediately affecting the framerate, leading to an annoying slow down as the value was decremented to 0.

I imagine this feature could be useful for any other slider where one would want to defer its update.

ZenithMDC committed:
Allow changing MSAA at runtime

ZenithMDC committed:
Update GUI texture filtering setting immediately

ZenithMDC commented:
This seemed too easy. Was there a reason this was being deferred until level load?

```diff
diff --git a/port/src/optionsmenu.c b/port/src/optionsmenu.c
index dfc12c97b..9e23290af 100644
--- a/port/src/optionsmenu.c
+++ b/port/src/optionsmenu.c
@@ -905,6 +905,7 @@ static MenuItemHandlerResult menuhandlerTexFilter2D(s32 operation, struct menuit
                return videoGetTextureFilter2D();
        case MENUOP_SET:
                videoSetTextureFilter2D(data->checkbox.value);
+               g_TexFilter2D = videoGetTextureFilter2D() ? G_TF_BILERP : G_TF_POINT;
                break;
        }

```

ZenithMDC committed:
Add Display FPS options to Extended Video Options

Also changed how the window title FPS works, which is now derived from
the difference in time between subsequent video frames.

ZenithMDC committed:
Fix mouse input not updating deferred slider while in dimmed mode

ZenithMDC committed:
Actually sync fullscreen and maximized states with the backend

ZenithMDC committed:
Rearrange the menu, clean up, and change some things

ZenithMDC committed:
Allow overriding g_TickRateDiv at runtime

ZenithMDC commented:
Updated the 1st post with new info.

ZenithMDC committed:
Call text_end (text0f153780) for correctness sake

fgsfdsfgs commented:
Sorry for taking 400 years to get back to this, but merged. Also made some small changes to it in ac4f3178, hopefully the commit message explains them well enough. If you have any questions, please let me know.

ZenithMDC commented:
No worries about taking forever to get around to it. It was a kind of big PR and you have other priorities, no doubt. I tested and everything seems fine. The changes you made are reasonable. Thanks for merging.

SilentException commented:
Testing on Linux, Steam Deck, some of the options (new and old) are not being saved in INI file.
For example: Anti-aliasing, Vsync, Framerate Limit, Texture Filtering, Detail Textures. Maybe there are others, I just quickly tested this.

fgsfdsfgs commented:
They are on my end. Are you quitting the game by hitting CTRL-C in the terminal or something?

SilentException commented:
> They are on my end. Are you quitting the game by hitting CTRL-C in the terminal or something?

Nope, clean exit from the in-game menu and confirming it.

Using `pd-x86_64-linux.tar.gz` from latest automatic build.

Just tried on Fedora 40, clear setup (without config folder or pd.ini). Same behavior.

SilentException commented:
> Do other settings save?

Yes, they do. 
From that menu specifically? Resolution, Uncap Tickrate, GE64-style Muzzle Flashes do get saved. GUI Texture Filtering does not (one more for the list above).

SilentException commented:
Figured it out. I was curious why I'm not getting any options after Video.AllowHiDpi so I assumed I'm hitting a limit of some sort. pd.ini had 288 which was very close to 255... So I removed all sections and empty lines and voila:
![image](https://github.com/user-attachments/assets/83367926-7ffd-42f6-acc5-e0842a83bac0)


So I'm hitting this limit, dunno why you aren't though.
https://github.com/fgsfdsfgs/perfect_dark/blob/8dbe6f2a6300d64cd3740f15ee9bf00cbedd3521/port/src/config.c#L13

fgsfdsfgs commented:
Very cool. I was probably hitting it on some other setting.

SilentException commented:
> Very cool. I was probably hitting it on some other setting. EDIT: Thanks, bumped the limit up.

Thanks, working good now! I've also submitted a small PR that will put a warning to the log when this limit is reached.

regionalradio commented:
Can this extended menu be added to the switch build too

fgsfdsfgs commented:
It should be in there. Download new builds from the releases.
